File: music-application/src/main/java/com/music/sale/application/category/port/outport/CategoryQueryPort.java
// Copyright (C) 2024 Your Name or Company
package com.music.sale.application.category.port.outport;

import com.music.sale.domain.category.Category;
import com.music.sale.domain.category.enums.CategoryType;
import java.util.List;
import org.springframework.stereotype.Component;

@Component
public interface CategoryQueryPort {
  Category getCategoryById(Long id);

  List<Category> findAll();

  Category findById(Long id);

  List<Category> findByType(CategoryType type);

  List<Category> findRootCategories();

  List<Category> findByParentId(Long parentId);

  Category save(Category category);

  void delete(Long id);
}

File: music-application/src/main/kotlin/com/music/sale/application/category/port/outport/CategoryQueryPort.kt
// Copyright (C) 2024 Your Name or Company
package com.music.sale.application.category.port.outport

import com.music.sale.domain.category.Category
import com.music.sale.domain.category.CategoryType
import org.springframework.stereotype.Component

@Component
interface CategoryQueryPort {
    fun getCategoryById(id: Long): Category

    fun findAll(): List<Category>

    fun findById(id: Long): Category?

    fun findByType(type: CategoryType): List<Category>

    fun findRootCategories(): List<Category>

    fun findByParentId(parentId: Long): List<Category>

    fun save(category: Category): Category

    fun delete(id: Long)
}
